### PR TITLE
Detector: Missing Inheritance

### PR DESCRIPTION
**Pull request by TilakMaddy**
Severity: Low

Qualifications: All solc versions

Logic:

```python
For every non abstract contract c which doesn't inherit from any other contract: 
        Check if there is  any abstract contract or interface  i  whose linearized base chain shares all the function selectors with c:
             Report i as missing inheritance of c 
```

When such a contract is found, report it

### CODE REVIEW

File: reports/ccip-functions-report.md [hunk 1]
```diff
@@ -2690,3 +2691,70 @@ Invoking `SSTORE`operations in loops may lead to Out-of-gas errors. Use a local
...
+	```solidity
+	contract FunctionsCoordinator is OCR2Base, IFunctionsCoordinator, FunctionsBilling {
+	```
+
+- Found in src/v0.8/functions/v1_3_0/FunctionsCoordinator.sol [Line: 13](../tests/ccip-contracts/contracts/src/v0.8/functions/v1_3_0/FunctionsCoordinator.sol#L13)
+
+	Consider implementing the interface - IFunctionsCoordinator
+	```solidity
+	contract FunctionsCoordinator is OCR2Base, IFunctionsCoordinator, FunctionsBilling {
+	```
```
Comment (TilakMaddy) [inline reports/ccip-functions-report.md:2748]: ooo Interesting case ...

@alexroan Help me here please. What shall we do in this case?

Here, there are 2 interfaces named IFunctionsCoordinator . That's why this has been caught 
Screenshot 2024-08-31 at 12 26 41 PM

File: reports/templegold-report.md [hunk 2]
```diff
@@ -8887,3 +8888,49 @@ Invoking `SSTORE`operations in loops may lead to Out-of-gas errors. Use a local
...
+	```solidity
+	contract TempleGoldAdmin is ITempleGoldAdmin, TempleElevatedAccess {
+	```
+
+- Found in contracts/v2/TempleDebtToken.sol [Line: 39](../tests/2024-07-templegold/protocol/contracts/v2/TempleDebtToken.sol#L39)
+
+	Consider implementing the interface - AMO_IAuraToken
+	```solidity
+	contract TempleDebtToken is ITempleDebtToken, TempleElevatedAccess {
+	```
```
Comment (TilakMaddy) [inline reports/templegold-report.md:8931]: Also give me your thoughts on this. Is it a useful suggestion ?

Comment (TilakMaddy) [reply reports/templegold-report.md:8931]: I am in the middle on this issue

**Comment by TilakMaddy** [comment]
Another doubt- 

If a contract A finds multiple interfaces that it could potentially inherit from,  what might we suggest? What's a good practice?

File: aderyn_core/src/detect/low/missing_inheritance.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        self.hints.clone()
+    }
+
+    fn title(&self) -> String {
+        String::from("Potentially missing inheritance for contract.")
+    }
+
+    fn description(&self) -> String {
+        String::from("There is an interface / abstract contract that is potentially missing (not included in) the inheritance of this contract. If that's not the case, consider using the same interface instead of defining multiple identical interfaces.")
+    }
```
Comment (TilakMaddy) [inline aderyn_core/src/detect/low/missing_inheritance.rs:151]: Do you agree with this @alexroan  (the last part) ?

Comment (TilakMaddy) [reply aderyn_core/src/detect/low/missing_inheritance.rs:151]: If not, we can adjust the logic to report missing inheritance only if it defines unique functions such that no other interface that is implemented, defines them.

File: reports/ccip-functions-report.md (hunk 1 above)
Comment (TilakMaddy) [reply reports/ccip-functions-report.md]: Or should we assume the re-used contract name detector catches that and therefore the dev has resolved it ?

File: aderyn_core/src/detect/low/missing_inheritance.rs (hunk 3 above)
Comment (TilakMaddy) [reply aderyn_core/src/detect/low/missing_inheritance.rs:151]: IDEA - I Will restrict this detector to run on contracts that have don't inherit from anywhere. This way we are respecting the developer's choice

File: reports/ccip-functions-report.md (hunk 1 above)
Comment (TilakMaddy) [reply reports/ccip-functions-report.md]: UPDATE: I'm going to restrict this detector to run on only those contracts that don't inherit from anything

File: reports/templegold-report.md (hunk 2 above)
Comment (TilakMaddy) [reply reports/templegold-report.md]: UPDATE: I'm going to restrict this detector to only run for those contracts that don't inherit from anything

File: aderyn_core/src/detect/low/missing_inheritance.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                    .is_some_and(|chain| chain.len() != 1)
+                {
+                    continue;
+                }
+                let missing_inheritances_vector =
+                    missing_inheritances.iter().cloned().collect::<Vec<_>>();
+                let missing_inheritaces_string = missing_inheritances_vector.join(", ");
+
+                let hint = format!(
+                    "Is this contract supposed to implement an interface? Consider extending one of the following: {}",
```
Comment (alexroan) [inline aderyn_core/src/detect/low/missing_inheritance.rs:134]: I updated this hint @TilakMaddy

**Pull request merged**